lib/pbench/agent/tool_data_sink.py

```diff
@@ -550,6 +610,117 @@ def terminate(self):
+    internal to the container ${pbench_run} path.  To resolve this, the entity
+    which creates the external pbench run directory creates a ".path" file in
+    that directory containing the full "external" path to the pbench run
+    directory. The Tool Data Sink uses that path to validate that external
+    benchmark_run_dir parameter values are valid.
+
+    This class implements the mechanism that allows the Tool Data Sink code to
+    handle that seamlessly.
+    """
+
+    def __init__(self, ext_benchmark_run_dir, int_pbench_run):
+        self._ext_benchmark_run_dir = Path(ext_benchmark_run_dir)
+        self._ext_pbench_run = self._ext_benchmark_run_dir.parent
+        self._int_pbench_run = Path(int_pbench_run)
+
+        # The Tool Data Sink could be running in a container. If so, then
+        # it'll be using the default benchmark run directory.  If the
+        # benchmark_run_dir parameter is valid, there will be a file
+        # called ".path" in the default benchmark run directory which will
+        # match.
+        #
+        # E.g.:
+        #  $ pbench_run="/home/<USER>/run-dir"
+        #  $ benchmark_run_dir="${pbench_run}/script_config_<date>"
+        #  $ cat ${pbench_run}/.path
+        #  /home/<USER>/run-dir
+        #  $ podman run --volume ${pbench_run}:/var/lib/pbench-agent \
+        #    pbench-agent-tool-data-sink bash
+        #  [ abcdefg /]$ cat /var/lib/pbench-agent/.path
+        #  /home/<USER>/run-dir
+        try:
+            benchmark_run_dir_lcl = self._ext_benchmark_run_dir.resolve(strict=True)
+        except Exception:
+            # Might be in a container; let's first construct the
+            # internal-to-the-container benchmark run directory.
+            benchmark_run_dir_lcl = (
+                self._int_pbench_run / self._ext_benchmark_run_dir.name
+            )
+            dot_path = self._int_pbench_run / ".path"
+            try:
+                dot_path_contents = dot_path.read_text().strip()
+            except Exception as exc:
+                # Failed to read ".path" contents, give up.
+                raise ToolDataSinkError(
+                    f"Run directory parameter, '{ext_benchmark_run_dir}', must"
+                    f" be an existing directory ('{self._ext_pbench_run}/"
+                    f".path' not found, '{exc}').",
+                )
+            else:
+                if dot_path_contents != str(self._ext_pbench_run):
+                    raise ToolDataSinkError(
+                        f"Run directory parameter, '{ext_benchmark_run_dir}',"
+                        " must be an existing directory (.path contents"
+                        f" mismatch, .path='{dot_path_contents}' !="
+                        f" '{self._ext_pbench_run}').",
+                    )
+        else:
+            # We can access the benchmark_run_dir directly, no need to
+            # consider contents of ".path" file.
+            pass
```

Comment:
The clause is documentation.  If we can read the external benchmark run directory, we are in the same file system name space as the entity that created the Tool Data Sink process.  If we can't read the external benchmark run directory, then we construct an internal run directory path to the same leaf name as the given external benchmark run directory.  If that directory exists, and it has a `.path` in which contains the same name as the external benchmark run directory we were given, then we know we have access to the same directory.  If we don't have that confirmation, we don't want to write any data anywhere, we must exit.

Reply:
The purpose of lines 671-700 appears to be to set the value of `benchmark_run_dir_lcl`, and that's done before consulting `.path`.  Everything from 679 onwards is testing the value of `.path`, but that value is not actually used -- by that point `benchmark_run_dir_lcl` has _already been set_.

So, it's not obvious how there would be any problem if we were to delete the last 20 lines of this block.  I understand that you're using the outermost `else` block for documentation purposes, but I don't think that it's worth taking up that space.  And, furthermore, I don't see why we need to worry about the contents of `.path`, since we don't seem to be basing anything on it

Reply:
So, you're trying to guard against the case where our guess at where to write the results from inside the container exists and is writable but is not actually the place where the user wants us to write?

Reply:
I chatted with Peter and he conveyed to me the level of pain and suffering that he experienced while developing the containerization.  This convinced me that the `.path` mechanism is worthwhile, in order to safeguard our users from a similar experience.  We can work on making it more friendly for the user to engage with, but it's worth keeping.